src/gt4py/storage/allocators.py

```diff
@@ -0,0 +1,399 @@
+        if not hasattr(self.ndarray, "__dlpack__"):
+            raise TypeError("Cannot export tensor buffer to DLPack.")
+        return self.ndarray.__dlpack__()
+
+    def __dlpack_device__(self) -> xtyping.DLPackDevice:
+        if not hasattr(self.ndarray, "__dlpack_device__"):
+            raise TypeError("Cannot extract DLPack device from tensor buffer.")
+        return self.ndarray.__dlpack_device__()
+
+
+class BufferAllocator(Protocol[_NDBufferT]):
+    """Protocol for buffer allocators."""
+
+    @property
+    def device_type(self) -> core_defs.DeviceType:
+        ...
+
+    def allocate(
+        self,
+        shape: Sequence[core_defs.IntegralScalar],
+        dtype: core_defs.DType[_ScalarT],
+        layout_map: BufferLayoutMap,
+        device: core_defs.Device,
+        byte_alignment: int,
+        aligned_index: Optional[Sequence[int]] = None,
+    ) -> TensorBuffer[_NDBufferT, _ScalarT]:
+        """
+        Allocate a TensorBuffer with the given shape, layout and alignment settings.
+
+        Args:
+            device: Device where the buffer is allocated.
+            dtype: Data type descriptor.
+            shape: Tensor dimensions.
+            layout_map: layout of the dimensions in the buffer.
+                layout_map[i] = j means that the i-th dimension of the tensor
+                corresponds to the j-th dimension of the buffer.
+            byte_alignment: Alignment (in bytes) of the first valid element.
+            aligned_index: N-dimensional index of the first aligned element.
+        """
+        ...
+
+
+@dataclasses.dataclass(frozen=True, init=False)
+class _BaseNDArrayBufferAllocator(abc.ABC, Generic[_NDBufferT]):
+    """Base class for buffer allocators using NumPy-like modules."""
+
+    def allocate(
+        self,
+        shape: Sequence[core_defs.IntegralScalar],
+        dtype: core_defs.DType[_ScalarT],
+        layout_map: BufferLayoutMap,
+        device: core_defs.Device,
+        byte_alignment: int,
+        aligned_index: Optional[Sequence[int]] = None,
+    ) -> TensorBuffer[_NDBufferT, _ScalarT]:
+        if not core_defs.is_valid_tensor_shape(shape):
+            raise ValueError(f"Invalid shape {shape}")
+        ndim = len(shape)
+        if len(layout_map) != ndim or not is_valid_layout_map(layout_map):
+            raise ValueError(f"Invalid layout_map {layout_map} for shape {shape}")
+
+        # Compute number of items inside an aligned block
+        item_size = dtype.byte_size
+        if math.gcd(byte_alignment, item_size) not in (byte_alignment, item_size):
+            raise ValueError(
+                f"Incompatible 'byte_alignment' ({byte_alignment}) and 'dtype' size ({item_size})"
+            )
+        items_per_aligned_block = (byte_alignment // item_size) or 1
+
+        # Compute the padding required in the contiguous dimension to get aligned blocks
+        dims_layout = [layout_map.index(i) for i in range(len(shape))]
+        padded_shape_lst = list(shape)
```

Comment:
This should be already enought for mypy

```suggestion
        padded_shape: Sequence[int] = list(shape)
```

Reply:
Doesn't work because `Sequence` is not mutable and `tuple` is not a `MutableSequence`...